cabal-install/Distribution/Client/CmdFreeze.hs

```diff
@@ -0,0 +1,166 @@
+{-# LANGUAGE NamedFieldPuns, RecordWildCards, CPP #-}
+
+-- | cabal-install CLI command: freeze
+--
+module Distribution.Client.CmdFreeze (
+    freezeAction,
+  ) where
+
+import Distribution.Client.ProjectPlanning
+         ( ElaboratedInstallPlan, rebuildInstallPlan )
+import Distribution.Client.ProjectConfig
+         ( ProjectConfig(..), ProjectConfigShared(..)
+         , commandLineFlagsToProjectConfig, writeProjectLocalFreezeConfig
+         , findProjectRoot )
+import Distribution.Client.ProjectPlanning.Types
+         ( ElaboratedConfiguredPackage(..) )
+import Distribution.Client.Targets
+         ( UserConstraint(..) )
+import Distribution.Solver.Types.ConstraintSource
+         ( ConstraintSource(..) )
+import Distribution.Client.DistDirLayout
+         ( defaultDistDirLayout, defaultCabalDirLayout )
+import Distribution.Client.Config
+         ( defaultCabalDir )
+import qualified Distribution.Client.InstallPlan as InstallPlan
+
+
+import Distribution.Package
+         ( PackageName, packageName, packageVersion )
+import Distribution.Version
+         ( VersionRange, thisVersion, unionVersionRanges )
+import Distribution.PackageDescription
+         ( FlagAssignment )
+import Distribution.Client.Setup
+         ( GlobalFlags, ConfigFlags(..), ConfigExFlags, InstallFlags )
+import Distribution.Simple.Setup
+         ( HaddockFlags, fromFlagOrDefault )
+import Distribution.Simple.Utils
+         ( die, notice )
+import Distribution.Verbosity
+         ( normal )
+
+#if !MIN_VERSION_base(4,8,0)
+import Data.Monoid (Monoid(mempty))
```

Comment:
If this is the only reason for CPP, we can avoid this by referring to `mempty` at least once qualified...